#
# Fixes menu location compatibility with WPML.
#
# Adds a menu location retrieval methods to the `Timber` class
# and adds a hook on the `theme_mod_nav_menu_locations` filter
# to translate menu IDs.
#
# This patch is a subset of the changeset of timber/timber#2733.
#
# See:
# - https://github.com/timber/timber/pull/2733
#
diff --git a/src/Factory/MenuFactory.php b/src/Factory/MenuFactory.php
index 34cbd8ef..efa35c4b 100644
--- a/src/Factory/MenuFactory.php
+++ b/src/Factory/MenuFactory.php
@@ -5,6 +5,7 @@ namespace Timber\Factory;
 use InvalidArgumentException;
 use Timber\CoreInterface;
 use Timber\Menu;
+use Timber\Timber;
 use WP_Term;
 
 /**
@@ -87,7 +88,7 @@ class MenuFactory
      */
     public function from_location(string $location, array $args = []): ?Menu
     {
-        $locations = \get_nav_menu_locations();
+        $locations = Timber::get_menu_locations();
         if (!isset($locations[$location])) {
             return null;
         }
@@ -212,7 +213,7 @@ class MenuFactory
          */
         $classmap = \apply_filters('timber/menu/classmap', []);
 
-        $location = $this->get_menu_location($term);
+        $location = Timber::get_menu_location($term);
 
         $class = $classmap[$location] ?? null;
 
@@ -251,18 +252,6 @@ class MenuFactory
         return $class;
     }
 
-    /**
-     * Get the menu location
-     *
-     * @param WP_Term $term
-     * @return string|null
-     */
-    protected function get_menu_location(WP_Term $term): ?string
-    {
-        $locations = \array_flip(\array_filter(\get_nav_menu_locations(), fn ($location) => \is_string($location) || \is_int($location)));
-        return $locations[$term->term_id] ?? null;
-    }
-
     /**
      * Build menu
      *
diff --git a/src/Integration/WpmlIntegration.php b/src/Integration/WpmlIntegration.php
index 6b19d45c..142fc4fc 100644
--- a/src/Integration/WpmlIntegration.php
+++ b/src/Integration/WpmlIntegration.php
@@ -14,10 +14,10 @@ class WpmlIntegration implements IntegrationInterface
 
     public function init(): void
     {
+        \add_filter('theme_mod_nav_menu_locations', [$this, 'theme_mod_nav_menu_locations'], 10, 1);
         \add_filter('timber/url_helper/file_system_to_url', [$this, 'file_system_to_url'], 10, 1);
         \add_filter('timber/url_helper/get_content_subdir/home_url', [$this, 'file_system_to_url'], 10, 1);
         \add_filter('timber/url_helper/url_to_file_system/path', [$this, 'file_system_to_url'], 10, 1);
-        \add_filter('timber/menu/id_from_location', [$this, 'menu_object_id_filter'], 10, 1);
         \add_filter('timber/menu/item_objects', [$this, 'menu_item_objects_filter'], 10, 1);
         \add_filter('timber/image_helper/_get_file_url/home_url', [$this, 'file_system_to_url'], 10, 1);
     }
@@ -30,11 +30,6 @@ class WpmlIntegration implements IntegrationInterface
         return $url;
     }
 
-    public function menu_object_id_filter($id)
-    {
-        return \wpml_object_id_filter($id, 'nav_menu');
-    }
-
     public function menu_item_objects_filter(array $items)
     {
         return \array_map(
@@ -42,4 +37,16 @@ class WpmlIntegration implements IntegrationInterface
             $items
         );
     }
+
+    public function theme_mod_nav_menu_locations($locations)
+    {
+        if (!\is_array($locations)) {
+            return $locations;
+        }
+
+        return \array_map(
+            fn ($id) => \wpml_object_id_filter($id, 'nav_menu'),
+            $locations
+        );
+    }
 }
diff --git a/src/Menu.php b/src/Menu.php
index b23d4170..eb8cd82a 100644
--- a/src/Menu.php
+++ b/src/Menu.php
@@ -293,10 +293,7 @@ class Menu extends CoreEntity
         }
 
         // Set theme location if available
-        $locations = \array_flip(\array_filter(\get_nav_menu_locations(), fn ($location) => \is_string($location) || \is_int($location)));
-
-        $this->theme_location = $locations[$term->term_id] ?? null;
-
+        $this->theme_location = Timber::get_menu_location($term);
         if ($this->theme_location) {
             $this->args->theme_location = $this->theme_location;
         }
diff --git a/src/Timber.php b/src/Timber.php
index 69e0a568..593d9fb7 100644
--- a/src/Timber.php
+++ b/src/Timber.php
@@ -1054,6 +1054,39 @@ class Timber
         return $menu;
     }
 
+    /**
+     * Get the navigation menu location assigned to the given menu.
+     *
+     * @param  WP_Term|int $term The menu to find; either a WP_Term object or a Term ID.
+     * @return string|null
+     */
+    public static function get_menu_location($term): ?string
+    {
+        if ($term instanceof WP_Term) {
+            $term_id = $term->term_id;
+        } elseif (\is_int($term)) {
+            $term_id = $term;
+        } else {
+            return null;
+        }
+
+        $locations = \array_flip(static::get_menu_locations());
+        return $locations[$term->term_id] ?? null;
+    }
+
+    /**
+     * Get the navigation menu locations with assigned menus.
+     *
+     * @return array<string, (int|string)>
+     */
+    public static function get_menu_locations(): array
+    {
+        return \array_filter(
+            \get_nav_menu_locations(),
+            fn ($location) => \is_string($location) || \is_int($location)
+        );
+    }
+
 
     /* Comment Retrieval
     ================================ */
